### PR TITLE
logging with log4j

### PR DESCRIPTION
**Pull request by danilexn**
To improve the interoperability with the new python bindings (via stimwrap) and simplify user experience, I replaced many 
`System.out.println` with a global `log4j` logger.

Also, there's a new progress bar class that can be used during pairwise alignment (similar to the GUI), for the console application. It uses the same `Consumer<Double>` paradigm as the GUI-based counterpart.

TODO:

- [ ] Add an argument to specify the logging level (or via environment variable)

### CODE REVIEW

**Comment by danilexn** [comment]
This also contains the changes from PR #7. This should be ok to merge when that one is in master.

File: src/main/java/align/GlobalOpt.java [hunk 1]
```diff
@@ -171,25 +175,25 @@ public static void iterativeGlobalOpt(
 			{
 				int unaligned = tc.preAlign().size();
 				if ( unaligned > 0 )
-					System.out.println( "(" + new Date( System.currentTimeMillis() ) + "): pre-aligned all tiles but " + unaligned );
+					logger.info( "(" + new Date( System.currentTimeMillis() ) + "): pre-aligned all tiles but " + unaligned );
```
Comment (minnerbe) [inline src/main/java/align/GlobalOpt.java:178]: Depending on the configuration, the logger might prepend a date string to log messages anyway. In this case, some logging messages can be simplified:  
```suggestion
					logger.info("pre-aligned all tiles except for {}", unaligned);
```

File: src/main/java/align/GlobalOpt.java [hunk 2]
```diff
@@ -171,25 +175,25 @@ public static void iterativeGlobalOpt(
...
+				logger.info( "(" + new Date( System.currentTimeMillis() ) + "):    Max Error: " + tc.getMaxError() + "px" );
 
 				// give some time for the output
 				try { Thread.sleep( 50 ); } catch ( Exception e) {}
 			}
 			catch (Exception e)
 			{
-				System.out.println( "Global optimization failed, please report this bug: " + e );
+				logger.error( "Global optimization failed, please report this bug: " + e );
 				e.printStackTrace();
```
Comment (minnerbe) [inline src/main/java/align/GlobalOpt.java:197]: I'm not sure, but I suspect that the logger reports the stack trace anyway (at least if called with the `... {}", e)` syntax). In that case, the `e.printStackTrace()` can be removed.

**Review by minnerbe — COMMENTED** [review]
The proper logging is great! I made a few comments regarding the logging, everything else was already reviewed in the other PR.
Again, I'll have a more in-depth look next week.

File: src/main/java/align/GlobalOpt.java (hunk 1 above)
Comment (danilexn) [reply src/main/java/align/GlobalOpt.java]: Addressed by [2e4f4fe](https://github.com/PreibischLab/STIM/pull/8/commits/2e4f4fe5a0f440cbea312b37bee7deafb38f6814)

File: src/main/java/align/GlobalOpt.java (hunk 2 above)
Comment (danilexn) [reply src/main/java/align/GlobalOpt.java]: Addressed by [782bb33](https://github.com/PreibischLab/STIM/pull/8/commits/782bb3317a3d19a3fb727831fa72bd31264ca98a)

**Comment by danilexn** [comment]
Since the commit #[773c166](https://github.com/PreibischLab/STIM/pull/8/commits/773c166bc476ae766f373caab6d2df558ce60a12) (which was tested and working w/ stimwrap), the following issues were introduced:

- [ ] transformation matrices are written as [0. 0. 0. 0. 0. 0.] in AnnData, after re-saving, after global alignment... Thus, I think it might be the function that writes the values to the file, and not the underlying objects having wrong values
- [ ] the `column-order` attribute from `obs` and `var` can be a binary string (formatted as a python list), and this is not handled properly now (an exception is thrown). Since the column-order is not critical for analysis with STIM, and since AnnData-python handles both cases anyway, I would recommend skipping ignoring column-order when the format is not String[]
- [ ] non-numeric values cannot be read from `obs` and `var`, since the readNumericalArray is being used, instead of the previous which supported dataframes. This is important as e.g., strings or categorical variables are often present.

**Comment by minnerbe** [comment]
Thanks for the feedback @danilexn !

> transformation matrices are written as [0. 0. 0. 0. 0. 0.] in AnnData, after re-saving, after global alignment... Thus, I think it might be the function that writes the values to the file, and not the underlying objects having wrong values

That's unfortunate. I encountered the same issues earlier yesterday when updating the N5 dependencies, but it seemed fixed after updating the anndata I/O code to my own external anndata library. I'll investigate this right now using the tutorials for stimwrap as a means to reproduce the error.

> the column-order attribute from obs and var can be a binary string (formatted as a python list), and this is not handled properly now (an exception is thrown). Since the column-order is not critical for analysis with STIM, and since AnnData-python handles both cases anyway, I would recommend skipping ignoring column-order when the format is not String[]

I remember troubleshooting a similar problem when writing my external anndata library, but this apparently didn't fix it. I'll wrap it in exception handling code so that this case is just ignored for now as you suggested. (The anndata standard requires the column-order attribute to be a string array, though.)

> non-numeric values cannot be read from obs and var, since the readNumericalArray is being used, instead of the previous which supported dataframes. This is important as e.g., strings or categorical variables are often present.

The currently used anndata library does pretty much the same thing as the previous code but in a more robust way (dataframe constraints are handled behind the scenes) and allows for Zarr and N5 backends in addition to the HDF5 one. In that sense, the previous code could also not handle Strings, since they are not covered by the generic type `<T extends NativeType<T> & RealType<T>>` because there is no string type that implements both constraints (in fact, `NativeType` is indented to only shadow primitive Java types as far as I understand). It would have simply thrown an exception when trying to read a non-numerical array. Anyway, even if we support reading this kind of data, STIM doesn't provide any means of processing it so far. I agree that the current state is not ideal, but I think we should delay working on this until we have an actual use case.

I'm strongly considering merging only part of this PR as you suggested for now and make a new one for the anndata updates. However, I would include 5 more commits (everything including [c9ef431](https://github.com/PreibischLab/STIM/pull/8/commits/c9ef431b057780284e2aa81d6c719d037453c2e5)), since the additional commits only add bugfixes.

**Comment by minnerbe** [comment]
Quick update: the first issue is solved by [02a0138](https://github.com/PreibischLab/STIM/pull/8/commits/02a01385bf9e0b10b3e63b46fda4b8acf6b1d7af), the second one by [901660a](https://github.com/PreibischLab/STIM/pull/8/commits/901660a12ecea6a4dabb48b1231d0c00ddb72cf4).

I'll thoroughly test the current state of the repo and merge it if I don't find anything.

**Comment by minnerbe** [comment]
It looks like [d470dc5](https://github.com/PreibischLab/STIM/pull/8/commits/d470dc5a46615bcb280c8e15a28d5daef008d12c) fixes the third issue.

After successfully testing stim + stimwrap on all available tutorials, I'm pretty confident that this PR can be merged in its entirety.

**Pull request merged**